sdk/core/Azure.Core/src/Pipeline/Internal/ResponseBodyPolicy.cs

```diff
@@ -75,7 +75,9 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
 
             if (message.BufferResponse)
             {
-                if (networkTimeout != Timeout.InfiniteTimeSpan)
+                // If cancellation is possible (whether due to network timeout or a user cancellation token being passed), then
+                // register callback to dispose the stream on cancellation.
+                if (networkTimeout != Timeout.InfiniteTimeSpan || oldToken.CanBeCanceled)
```

Comment:
Can you share why this fixes the issue (for my edification) ?  Is it that now that we've registered the callback, the stream will be disposed when the CT is cancelled, so it won't be disposed during read?

Reply:
There are two issues being fixed here:
1. We weren't translating NotSupportedException into TaskCanceledException
2. We were only registering the callback if the network timeout was not infinite even though the point of this code is to dispose the stream on cancellation.

Reply:
How did the Dispose concurrent with Read issue work?  And how does this prevent that?

Reply:
Similar to how we translate ObjectDisposedExceptions when the cancellation token is cancelled, we are going to translate NotSupportedExceptions into TaskCanceledExceptions when the user cancels. So we are not actually preventing the race condition, but rather recovering from it in the same way that we do for ObjectDisposedException. The reason we can't prevent these is we intentionally dispose the stream on cancellation to make a best effort attempt to get the processing to stop as soon as the user cancels.

Reply:
Correct - we throw what would be the expected exception under this scenario because the cancellation token has been cancelled.